(SERVER-2845) Serve file_content for bolt projects

pcarlisle opened the pull request:
This adds support to the v3/file_content endpoint for serving files out
of projects. This is used by passing the query parameter `project`
instead of `environment`. This is implemented entirely in clojure,
unlike environment file serving support which still uses ruby.

pcarlisle commented on src/clj/puppetlabs/services/master/master_core.clj:
```diff
@@ -873,6 +873,127 @@
+  (case mount
+    "modules" "files"
+    ;; default to itself
+    mount))
```
I probably need to be explicit about which are allowed, i think right now you could probably request out of any subdirectory because of this

mwaggett replied:
+1

pcarlisle replied:
Nevermind, this is already handled through the routing table and the error is returned in `handle-project-file-content`

mwaggett commented on src/clj/puppetlabs/services/master/master_core.clj:
```diff
@@ -873,6 +873,127 @@
+  "Handle a file_content request for a bolt project."
+  [bolt-project-dir request]
+  (let [project (get-in request [:params "project"])
+        version (get-in request [:params "version"])
```
Where does `version` come from?

pcarlisle replied:
leftover line from an alternate implementation where i treated the different components of the project name independently (which is still something we might choose to do later, not hard to change)

mwaggett commented on src/clj/puppetlabs/services/master/master_core.clj:
```diff
@@ -873,6 +873,127 @@
+      (fs/exists? (str dir "/Boltdir"))))
+
+(defn find-bolt-project
+  [project-dir project-name]
+  (->> project-dir
+       io/file
+       .list
+       seq
+       (filter #(is-bolt-project? (str project-dir "/" %)))
+       (some #{project-name})))
+
+(defn get-project-root
+  [project-dir project-name]
```
can `project-dir` be `bolt-project-dir`, to match what's used below? or maybe something like `bolt-projects-dir` might be clearer, to distinguish from what you've been calling `project-root`? but really it's the consistency I care about.

mwaggett commented on src/clj/puppetlabs/services/master/master_core.clj:
```diff
@@ -873,6 +873,127 @@
+  (if file
+    (-> file
+        str
+        rr/file-response
```
Is this what actually reads the file content? I was poking around and it didn't totally look like it? but I can't find where else that might be happening..?

pcarlisle replied:
ring automatically handles a File being returned in the body

mwaggett replied:
Can you explain how it does that? I'm looking at https://github.com/ring-clojure/ring/blob/master/ring-core/src/ring/util/response.clj and not seeing what actually gets the content of the file..

pcarlisle replied:
https://github.com/ring-clojure/ring/blob/master/ring-core/src/ring/core/protocols.clj#L37

mwaggett replied:
oh god. thanks.

mwaggett commented on src/clj/puppetlabs/services/master/master_service.clj:
```diff
@@ -100,6 +100,7 @@
          certname (get-in config [:puppetserver :certname])
          localcacert (get-in config [:puppetserver :localcacert])
          puppet-version (get-in config [:puppetserver :puppet-version])
+         bolt-project-dir (get-in config [:bolt :project-dir])
```
¯\_(ツ)_/¯ I like using plurals for things that will contain many of the thing, e.g. projects. so `:projects-dir`. I don't know if now is when we have to set these in stone, but figured I'd offer my opinion 😝

mwaggett commented on test/unit/puppetlabs/services/master/master_core_test.clj:
```diff
@@ -439,3 +441,51 @@
+               (get-project-root bolt-project-dir "embedded")))))
+
+    (testing "finding a file in a project"
+      (testing "in the modules mount"
```
should we test the tasks mount too?

pcarlisle replied:
added this in the service test

mwaggett commented on test/integration/puppetlabs/services/master/master_service_test.clj:
```diff
@@ -766,3 +766,26 @@
             (is (= 503 status-code))
             (is (<= 0 retry-after 1800))))))))
+
+(deftest ^:integration project-file-content
```
Was there a plan to do any testing with the `named` project?
There seem to be a handful of unused test files (though maybe some are just necessary for structuring a project correctly/realistically)?

pcarlisle replied:
Originally I had a question of whether we ever needed to identify a project by it's name from the yaml file when that's different from the directory name. I think that we should just avoid this though, since we're going to write how deployment works.

mwaggett replied:
hmm, I think it makes sense to avoid that here. but I'm wondering where in the workflow that needs to happen.. Do we allow a user to deploy a project using the name from the yaml file (as opposed the name of the directory/repo)? If so, will the front end handle that or do we need to handle that in code-manager? 

If we're not gonna deal with it here, I think we should remove the associated test files.

mwaggett commented on documentation/puppet-api/v3/file_content.markdown:
```diff
@@ -0,0 +1,32 @@
+in the response body. An unsuccessful request returns an error response
+code with a `text/plain` Content-Type header:
+
+-   400: returned when neither the environment nor the project query parameters are provided, or when both project and environment are provided. 
+-   404: returned when requesting a file that is not within a module's `files` or `tasks`
```
Should we note the error responses for when the `project` isn't found or the file path is missing from the request?
I'm comparing to the existing docs for this endpoint: https://puppet.com/docs/puppet/6.17/http_api/http_file_content.html

pcarlisle replied:
they're all 404 but i added something to this line to mention that

pcarlisle commented:
addressed all feedback up to this point, added some more docstrings, removed a function that I didn't realize was unused and its tests, expanded the service test a bit

mwaggett commented on src/clj/puppetlabs/services/master/master_core.clj:
```diff
@@ -873,6 +873,121 @@
+      (str projects-dir "/" project-name))))
+
+(defn dirs-in-project-modulepath
+  "List all directories in a bolt project's modulepath. Returns a sequence of
```
Is it really a sequence?? That's the one where order matters, right? I didn't think that's what was returned here..

mwaggett commented on test/unit/puppetlabs/services/master/master_core_test.clj:
```diff
@@ -439,3 +441,33 @@
+
+    (testing "finding a file in a project"
+      (testing "in the modules mount"
+        (is (= (fs/file (str bolt-projects-dir "/local/modules/helpers/files/marco.sh" ))
+               (fs/file (find-project-file bolt-projects-dir "local" "modules" "helpers" "marco.sh"))))
+        (is (= (fs/file (str bolt-projects-dir "/local/site-modules/utilities/files/etc/greeting" ))
+               (fs/file (find-project-file bolt-projects-dir "local" "modules" "utilities" "etc/greeting")))))
```
probably should just compare strings here, instead of wrapping both sides in `fs/file`

donoghuc commented on src/clj/puppetlabs/services/master/master_core.clj:
```diff
@@ -873,6 +873,121 @@
 
+(defn is-bolt-project?
+  [dir]
+  (or (fs/exists? (str dir "/bolt-project.yaml"))
```
I think the *only* configuration we can support is `bolt-project.yaml` and `Boltdir/bolt-project.yaml`. We should not care about bolt.yaml and in the case we have a Boltdir, we should make sure that there is a `Boltdir/bolt-project.yaml`.
